File: app/src/main/java/com/vocaby/app/NotificationReceiver.java
package com.vocaby.app;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;

import androidx.core.app.NotificationCompat;

import com.vocaby.app.database.DatabaseManager;

import java.util.List;

public class NotificationReceiver extends BroadcastReceiver {
    private static final String CHANNEL_ID = "VOCABY_CHANNEL";
    private static final String NOTIFICATION_PICKS = "notificationPicks";
    private static final String PREV_PICK = "notificationPrevPick";

    @Override
    public void onReceive(Context context, Intent intent) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        SharedPreferences sharedPreferences = context.getSharedPreferences(NOTIFICATION_PICKS, Context.MODE_PRIVATE);
        String prevWord = sharedPreferences.getString(PREV_PICK, null);
        DataManager dataManager = DataManager.getInstance(context);
        List<String> saves = dataManager.getSaves();
        WordPickerService wordPickerService = new WordPickerService(saves, context);
        int index = wordPickerService.getRandomWordFromSaves(prevWord);
        DatabaseManager databaseManager = DatabaseManager.getInstance(context);
        databaseManager.openDatabase();

        Word wordData = databaseManager.getWordData(saves.get(index));
        databaseManager.closeDatabase();
        String word;
        String pos;
        String message;

        if(saves.isEmpty()) {
            word = "No Saved Words";
            message = "Save words in the app to display in the notification";
        } else {
            word = wordData.getWord();
            pos = wordData.getAllowedPos()[0];
            message = wordData.getDefinitions(pos)[0];

            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString(PREV_PICK, word);
            editor.apply();
        }

        createNotificationChannel(notificationManager);

        Intent resultIntent = new Intent(context, MainActivity.class);
        resultIntent.putExtra("com.vocaby.app.openAndSearch", word);
        resultIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
        PendingIntent resultPendingIntent = PendingIntent.getActivity(context, 0, resultIntent, PendingIntent.FLAG_CANCEL_CURRENT);


        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_notification_white)
                .setColor(context.getColor(R.color.colorPrimary))
                .setContentTitle(word.toUpperCase())
                .setContentIntent(resultPendingIntent)
                .setStyle(new NotificationCompat.BigTextStyle()
                        .bigText(message))
                .setContentText(message);
        notificationManager.notify(777, builder.build());
    }

    private void createNotificationChannel(NotificationManager notificationManager) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "Vocaby Notification";
            String description = "Vocaby Notification";
            int importance = NotificationManager.IMPORTANCE_LOW;
            NotificationChannel channel = new NotificationChannel(CHANNEL_ID, name, importance);
            channel.setDescription(description);
            notificationManager.createNotificationChannel(channel);
        }
    }
}

File: app/src/main/java/com/vocaby/application/feature_profile/presentation/setting/receivers/NotificationReceiver.kt
package com.vocaby.application.feature_profile.presentation.setting.receivers

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.app.NotificationCompat
import com.vocaby.application.R
import com.vocaby.application.core.presentation.MainActivity
import com.vocaby.application.core.util.Generators.generateRandomInt
import com.vocaby.application.feature_dictionary.domain.model.EntryModel
import com.vocaby.application.feature_dictionary.domain.repository.DictionaryRepository
import com.vocaby.application.feature_dictionary_custom.domain.repository.CustomDictionaryRepository
import com.vocaby.application.feature_profile.domain.repository.UserRepository
import com.vocaby.application.feature_save.domain.model.SaveCollectionModel
import com.vocaby.application.feature_save.domain.repository.SaveRepository
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class NotificationReceiver : BroadcastReceiver() {
    @Inject
    lateinit var dictionaryRepository: DictionaryRepository
    @Inject
    lateinit var customDictionaryRepository: CustomDictionaryRepository
    @Inject
    lateinit var saveRepository: SaveRepository
    @Inject
    lateinit var userRepository: UserRepository

    override fun onReceive(context: Context, intent: Intent) {
        val notificationManager =
            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val scope = CoroutineScope(Dispatchers.Main.immediate)
        val sp = context.getSharedPreferences("SAVES", Context.MODE_PRIVATE)

        scope.launch(Dispatchers.Default) {
            val userId = userRepository.getUser()
            val settings = userRepository.settingsFlow.first()
            val selectedCollectionId = settings.notificationCollectionId
            val collection = saveRepository.getSaveCollectionWithId(selectedCollectionId)
            var isFromCollection = true
            var saves = saveRepository.getCollectionItems(selectedCollectionId).map { it.entry }
            if (collection == null || selectedCollectionId < 1) {
                isFromCollection = false
                saves = saveRepository.getAllSavedEntriesFlow(userId).first()
            }

            if (saves.isEmpty()) {
                val editor = sp.edit()
                editor.putString("NOTIF_PREV_SELECT", "")
                editor.apply()

                val title = "No Saved Words"
                val message = "Save words in the app to display in the notification"
                createNotification(
                    context,
                    notificationManager,
                    title,
                    isFromCollection,
                    collection,
                    message
                )
            } else {
                var index = generateRandomInt(0, saves.size-1)
                val prevWord = sp.getString("NOTIF_PREV_SELECT", "")

                while (saves.size != 1 && saves[index] == prevWord) {
                    index = generateRandomInt(0, saves.size-1)
                }

                val entry = saves[index]
                val entryModel: EntryModel? =
                    customDictionaryRepository.getUserEntryData(userId, entry) ?:
                    dictionaryRepository.getEntryDataFromDatabase(entry)

                var message = "No definition found"

                entryModel?.let { model ->
                    model.firstGroup?.let {
                        message = it.definitionData[0].definition
                    }
                }

                sp.edit().putString("NOTIF_PREV_SELECT", entry).apply()
                createNotification(
                    context,
                    notificationManager,
                    entry,
                    isFromCollection,
                    collection,
                    message
                )
            }
        }
    }

    private fun createNotification(
        context: Context,
        notificationManager: NotificationManager,
        title: String,
        isFromCollection: Boolean,
        collection: SaveCollectionModel?,
        message: String
    ) {
        createNotificationChannel(notificationManager)
        val builder = NotificationCompat.Builder(context, CHANNEL_ID)
        val resultIntent = Intent(context, MainActivity::class.java)
        resultIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
        val resultPendingIntent = PendingIntent.getActivity(
            context, 0,
            resultIntent, PendingIntent.FLAG_IMMUTABLE
        )

        if (isFromCollection && collection != null) {
            builder.setSubText(collection.collectionName)
        }

        builder.setSmallIcon(R.drawable.ic_notification_white)
            .setColor(context.getColor(R.color.colorPrimary))
            .setContentTitle(title)
            .setContentIntent(resultPendingIntent)
            .setStyle(
                NotificationCompat.BigTextStyle()
                    .bigText(message)
            )
            .setContentText(message)

        notificationManager.notify(313, builder.build())
    }

    private fun createNotificationChannel(notificationManager: NotificationManager) {
        val name: CharSequence = "Vocaby Notification"
        val description = "Vocaby Notification"
        val importance = NotificationManager.IMPORTANCE_LOW
        val channel = NotificationChannel(CHANNEL_ID, name, importance)
        channel.description = description
        notificationManager.createNotificationChannel(channel)
    }

    companion object {
        private const val CHANNEL_ID = "VOCABY_CHANNEL"
    }
}